typedData/typedData.go

```diff
@@ -902,34 +915,128 @@ func (domain *Domain) UnmarshalJSON(data []byte) error {
 		return err
 	}
 
-	revision, err := getField("revision")
-	if err != nil {
-		revision = "0"
+	// Custom logic to handle the `revision` field,
+	// used to marshal the Revision exactly as it is in the original JSON.
+	rawRevision, ok := dec["revision"]
+	if !ok {
+		rawRevision = "0"
 	}
-	numRevision, err := strconv.ParseUint(revision, 10, 8)
-	if err != nil {
-		return err
+
+	var numRevision uint64
+	switch revision := rawRevision.(type) {
+	case string:
+		domain.HasStringRevision = true
+		numRevision, err = strconv.ParseUint(revision, 10, 8)
+		if err != nil {
+			return err
+		}
+	case float64:
+		domain.HasStringRevision = false
+		numRevision = uint64(revision)
+	}
+
+	// Custom logic to handle the `chainId` field,
+	// used to marshal the ChainId exactly as it is in the original JSON.
+	rawChainId, ok := dec["chainId"]
+	if !ok {
+		err = errors.New("error getting the value of 'chainId' from 'domain' struct")
 	}
 
-	chainId, err := getField("chainId")
 	if err != nil {
 		if numRevision == 1 {
 			return err
 		}
-		var err2 error
+
+		// `chain_id` was also used in the past, so we check for it if the `chainId` field is not found
 		// ref: https://community.starknet.io/t/signing-transactions-and-off-chain-messages/66
-		chainId, err2 = getField("chain_id")
-		if err2 != nil {
+		rawChainId, ok = dec["chain_id"]
+		if !ok {
+			err2 := errors.New("error getting the value of 'chain_id' from 'domain' struct")
+
 			return fmt.Errorf("%w: %w", err, err2)
 		}
+		domain.hasOldChainIdName = true
+	}
+
+	switch rawChainId.(type) {
+	case string:
+		domain.hasStringChainId = true
+	case float64:
+		domain.hasStringChainId = false
 	}
+	chainId := fmt.Sprintf("%v", rawChainId)
 
+	// Final step
 	*domain = Domain{
 		Name:     name,
 		Version:  version,
 		ChainId:  chainId,
 		Revision: uint8(numRevision),
+
+		hasStringChainId:  domain.hasStringChainId,
+		hasOldChainIdName: domain.hasOldChainIdName,
+		HasStringRevision: domain.HasStringRevision,
 	}
 
 	return nil
 }
+
+// MarshalJSON implements the json.Marshaler interface for Domain.
+// Some logic was added to marshal the `Domain` exactly as it is in the original JSON.
+func (domain Domain) MarshalJSON() ([]byte, error) {
```

Comment:
Receive `Domain` by pointer instead of by value. This will cause a very heavy copy. I am surprised there is no linter error here? Do you have golangci-linter as part of the actions?